### PR TITLE
Add default HTTP User-Agent for specific system

### LINKED ISSUE #2408: Can not load image with Url whitch contains ","
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios 
 Platform Version        | 11.0 / 10.0
 SDWebImage Version      | 4.4.1
 Integration Method      | cocoapods
 Xcode Version           |  Xcode 9.4
 Repro rate              | all the time (100%)
 Repro with our demo prj | it happen with your demo project

### Issue Description and Steps
Issue: Can not load image with Url whitch contains ","
sample url: http://img2.imgtn.bdimg.com/it/u=4158051125,1333702261&fm=21&gp=0.jpg

### PR DESCRIPTION
**Pull request by zhongwuzw**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: ...

### Pull Request Description

Let's tell the server what we are. Server maybe forbidden us if he think it's not the valid request.

Don't needs to reinventing the wheel, borrow snippet from [AFNetWorking](https://github.com/AFNetworking/AFNetworking/blob/a24b4896bb6e62f73b4dbc7ccf9f11f1a8720f11/AFNetworking/AFURLRequestSerialization.m#L220-L238).

Fixed #2408 .

### CODE REVIEW

**Comment by dreampiggy** [comment]
Use [headersFilter](https://github.com/rs/SDWebImage/blob/master/SDWebImage/SDWebImageDownloader.h#L169-L175) in 4.x instead.

For 5.x, use [requestModifier](https://github.com/rs/SDWebImage/blob/5.x/SDWebImage/SDWebImageDownloader.h#L133-L139) instead.

We should not bind complicated business code to a common library. If the user desired to use something, we just provide a abstract protocol or way to let them provide the information.

**Comment by zhongwuzw** [comment]
@dreampiggy It has no any business code, it just add the device and system version. just like all common user agent. you can see `UIWebView`/`WKWebView` or `Safari`.

Like below, you can see how Apple uses the user-agent to identify the device.

image

**Comment by bpoplauschi** [comment]
@zhongwuzw I agree we need to set the `User-Agent`. I would also consider the comment from @dreampiggy and actually make use of our existing features to do so. 
Just that we need this `User-Agent` configured properly for all our users. We can't expect them to do it. We only offer ways to alter this behavior, if they need that.

**Comment by zhongwuzw** [comment]
@bpoplauschi the propose @dreampiggy suggested is used for custom for each request, so any opinion about this `PR`?

**Comment by bpoplauschi** [comment]
I've looked at this PR again. Guys, I think an API like `requestModifier` is too advanced for the users that simply use our `UIImageView` (or similar) categories. I see AFNetworking also sets the `User-Agent` in the same way, so I say we do the same.
This is not a case where some users want some custom behaviour. From my point of view, this is the `User-Agent` set for each request of all our users, so we need to set it properly.
The only feature I say we could add (but it's just nice to have) is another field in the `DownloaderConfig` that allows passing in the `User-Agent`.

**Comment by zhongwuzw** [comment]
@bpoplauschi Emm 🤔 , `User-Agent` is only a field of `Headers`, user can override these values by `(void)setValue:(nullable NSString *)value forHTTPHeaderField:(nullable NSString *)field` of `SDWebImageDownloader or `requestModifier`, I think it's maybe enough. Would you think we must add the `Boolean` in `DownloaderConfig`?

File: SDWebImage/SDWebImageDownloader.m [hunk 1]
```diff
@@ -94,11 +94,32 @@ - (nonnull instancetype)initWithSessionConfiguration:(nullable NSURLSessionConfi
         _downloadQueue.maxConcurrentOperationCount = 6;
         _downloadQueue.name = @"com.hackemist.SDWebImageDownloader";
         _URLOperations = [NSMutableDictionary new];
+        SDHTTPHeadersMutableDictionary *headerDictionary = [SDHTTPHeadersMutableDictionary dictionary];
+        NSString *userAgent = nil;
+#if SD_IOS
```
Comment (dreampiggy) [inline SDWebImage/SDWebImageDownloader.m]: This seems does not handle tvOS, so it's better to use `SD_UIKIT` instead.

**Comment by dreampiggy** [comment]
@zhongwuzw Don't need to add one extra config in `SDWebImageDownloadConfig` or somewhere. If you want to add this, it shuold be one built-in HTTP headers. Use can use current APIs to change or override it. It's not suitable to add any property just for `User-Agent` this specify headers.

**Comment by bpoplauschi** [comment]
You are right guys, I forgot about `setValue:forHTTPHeader:`. That is sufficient for the users to customize. So we just need to calculate the `User-Agent` for the default case, which is most of the users (that will never deal with that). Basically this PR + the minor change request by @dreampiggy

File: SDWebImage/SDWebImageDownloader.m (hunk 1 above)
Comment (zhongwuzw) [reply SDWebImage/SDWebImageDownloader.m]: Nice catch.

**Comment by bpoplauschi** [comment]
I see no reason to keep this un-merged. Good job @zhongwuzw

**Pull request merged**